File: algorithm/C++/0322-coin_change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<int> table(amount+1, 10001);
        table[0] = 0;
        for (int i=1; i<=amount; ++i){
            for(int j=0; j<n; ++j){
                if(coins[j]<=i){
                    table[i] = min(1+table[i-coins[j]], table[i]);
                }
            }
        }
        return table[amount] == 10001 ? -1: table[amount];
    }
};

File: algorithm/Go/0322-Coin_Change.go
func coinChange(coins []int, amount int) int {
    table := make([]int, amount + 1)
    for i:=0; i<=amount; i++ {
        table[i] = math.MaxInt
    }
    table[0] = 0
    for i:=0; i<=amount; i++ {
        for j:=0; j<len(coins); j++ {
            if i >= coins[j] {
                if table[i - coins[j]] != math.MaxInt {
                    if table[i] > table[i - coins[j]] + 1 {
                        table[i] = table[i - coins[j]] + 1
                    }
                }
            }
        }
    }
    if table[amount] == math.MaxInt {
        return -1
    }
    return table[amount]
}
// Time : O(nk)
// Space : O(n)